File: backTest.py
# -*- coding: utf-8 -*-
"""
Created on Tue Sep 29 09:49:13 2020

This function perform a backtest of portfolio of risky securities.
The portfolio can be constructed in several different ways:
(1) Equal weighting (equal)
(2) Equal vol-weighting (equalvol)
(3) Mean-Variance optimization (mv)
(4) black litterman optimization (bl)

Inputs: 
    param - structure that specifies the backtest to run
    Data  - dataset (factor returns and covariances, security names, and
            rebalancing dates)
Outputs:
    output_struct - structure with extensive outputs
Notes:
1. We assume the first portfolio is all in cash
2. Portfolio weights are stored in their cash allocation (dollars) and indexed such that W(t) align with Rets(t):
(a) Rets(t) corresponds to the security return over the time period (t-1,t]
(b) W(t) corresponds to the portfolio weights held over the time period (t-1,t]
3. The value of the portfolio is calculated at the end of each time
period. In other words, if the time period is (t-1,t] the portfolio value
is calculated at time t.

@author: Shaolun Du
@contact: Shaolun.du@gmail.com
"""
import numpy as np
from cvxopt import solvers
from cvxopt import matrix
# global silent the solvers
solvers.options['show_progress'] = False

def backTest(param, Data, if_debug):
    # read info from Dataset
    secnames   = Data['secnames']
    rets       = Data['rets']
    dates      = Data['dates']
    COV        = Data['COV']
    ER         = Data['ER']
    rebaldates = Data['rebaldates']
    nAssets    = Data['nAssets']
    # Define variables and allocate space
    T              = len(dates)
    cash           = np.zeros(T)
    tcosts         = np.zeros(T)
    turnover       = np.zeros(T)
    portVal        = np.zeros(T)
    portRet        = np.zeros(T)
    retdates       = np.zeros(T)
    W              = np.zeros((nAssets,T))
    RiskContr      = np.zeros((nAssets,T))
    securityRet    = np.zeros((nAssets,T))
    RiskProj       = np.zeros(T)
    SigStale       = np.zeros((nAssets,nAssets))
    cash[0]        = param['capital']
    portVal[0]     = cash[0]
    portValCurrent = portVal[0]
    wCurrent       = W[:,0]
    # Now move forward in time and rebalalance when needed
    for t in range(1,T):
        # Error checking if current port value below zero
        if portVal[t]<0:
            raise Exception('Port val less than zero!{}\n'.format(portVal[t]))
        date_t = dates[t]
        if if_debug:
            print("Current Time:{}".format(date_t))
            print("Current weights:{}".format(wCurrent))
        # Covariance matrix 
        Sig = COV[date_t]
        mu  = ER[date_t]
        # Determine if we need to rebalance
        if_rebal = True if date_t in rebaldates else False
        if if_rebal:  # We need to rebalance
            if if_debug:
                print('Rebalancing...')
            SigStale = Sig
            if param['PortConstr'] == 'equal':
                wNew = portValCurrent*(np.ones(nAssets)/nAssets)
            if param['PortConstr'] == 'equalvol':
                vols       = np.sqrt(np.diagonal(Sig))
                volsinv    = 1/vols
                volsinvsum = sum(volsinv)
                wNew       = portValCurrent*(volsinv/volsinvsum)
            if param['PortConstr'] == 'mv':
                # Here lambda stands for 2*lambda
                # Long only weights
                lambda2 = 8
                P = matrix(Sig*lambda2)
                q = matrix(-mu.T)
                G = matrix(-np.eye(nAssets))
                h = matrix(np.zeros(nAssets))
                A = matrix(np.ones((1,nAssets)))
                b = matrix(np.ones(1))
                wNew = portValCurrent*np.array(solvers.qp(P,q,G,h,A,b)['x']).reshape((nAssets,))
            if param['PortConstr'] == 'bl':
                # black litterman model
                # cov= pca(cov)
                # er = avg(market equal,momentum)
                lambda2 = 8
                P = matrix(Sig*lambda2)
                q = matrix(-mu.T)
                G = matrix(-np.eye(nAssets))
                h = matrix(np.zeros(nAssets))
                A = matrix(np.ones((1,nAssets)))
                b = matrix(np.ones(1))
                wNew = portValCurrent*np.array(solvers.qp(P,q,G,h,A,b)['x']).reshape((nAssets,))
            
        else:
            # We don't need to rebalance
            wNew = wCurrent
        wTsigw           = np.dot(np.dot(wNew,Sig),wNew)
        wTsigwStale      = np.dot(np.dot(wNew,SigStale),wNew)
        W[:,t]           = wNew
        RiskProj[t]      = np.sqrt(wTsigw)/sum(wNew)
        RiskContr[:,t]   = np.dot(wNew,np.dot(Sig,wNew))/wTsigwStale
        trades           = wNew - wCurrent
        turnover[t]      = sum(abs(trades))/portValCurrent
        # Calculate portfolio at the end of the period
        # (If needed in the futute, add cash position here)
        wCurrent         = np.multiply(wNew,1+rets.iloc[t].to_numpy())
        securityRet[:,t] = np.dot(wNew,rets.iloc[t].to_numpy().T)
        portValCurrent   = sum(wCurrent)
        portVal[t]       = portValCurrent
        portRet[t]       = (portVal[t]-portVal[t-1])/portVal[t-1]
        retdates[t]      = date_t
        
    
    # Populate output structure
    output_struct                  = {}
    output_struct['portVal']       = portVal
    output_struct['dates']         = dates
    output_struct['W']             = W
    output_struct['RiskContrProj'] = RiskContr
    output_struct['RiskProj']      = RiskProj
    output_struct['cash']          = cash
    output_struct['turnover']      = turnover
    output_struct['secnames']      = secnames
    output_struct['securityRet']   = securityRet
    output_struct['portRet']       = portRet
    output_struct['retdates']      = retdates    
    output_struct['PortConstr']    = param['PortConstr']  
    return output_struct

File: MatlabCode/BackTest.m
function output_struct = Backtest(param,Data,if_debug)
% This function perform a backtest of portfolio of risky securities.
% The portfolio can be constructed in several different ways:
% (1) Equal weighting (equal),
% (2) Equal vol-weighting (equalvol),
% (3) Global minimum variance portfolio, and  (FOR YOU TO IMPLEMENT)
% (4) Mean-Variance optimization (mv)         (FOR YOU TO IMPLEMENT)
% 
% Inputs: 
%     param - structure that specifies the backtest to run
%     Data  - dataset (factor returns and covariances, security names, and
%             rebalancing dates)
%
% Outputs:
%     output_struct - structure with extensive outputs
%
% Notes:
% 1. We assume the first portfolio is all in cash
% 2. Portfolio weights are stored in their cash allocation (dollars) and indexed such that W(t) align with Rets(t):
% (a) Rets(t) corresponds to the security return over the time period (t-1,t]
% (b) W(t) corresponds to the portfolio weights held over the time period (t-1,t]
% 3. The value of the portfolio is calculated at the end of each time
% period. In other words, if the time period is (t-1,t] the portfolio value
% is calculated at time t.

% Petter Kolm Copyright 2012-2016. 
%
% This file cannot be redistributed without the explicit permission
% of Petter Kolm, Petter.Kolm@nyu.edu.
%
% Please report any bugs to Petter.Kolm@nyu.edu


secnames = Data.secnames;
Rets = Data.Rets;
dates = Data.dates; 
COV = Data.COV;
ER = Data.ER;
rebaldates = Data.rebaldates;
nAssets = Data.param_Data.nAssets;

% Define variables anbd allocate space
T = length(dates);
cash = zeros(T,1);
tcosts = zeros(T,1);
turnover = zeros(T,1);
portVal = zeros(T,1);
portRet = zeros(T-1,1);
retdates = zeros(T-1,1);
W = zeros(nAssets,T);
RiskContr = zeros(nAssets,T-1);
securityRet = zeros(nAssets,T-1);
RiskProj = zeros(T-1,1);

% At t = 1
% . . . Initial portfolio value (i.e. value at starttime) and risk factor
% allocations
cash(1) = param.capital;
portVal(1) = cash(1);

portValCurrent = portVal(1);
wCurrent = W(:,1);
% Now move forward in time and rebalalance when needed
for t = 2:T,
    if if_debug
        t
    end
    % Covariance matrix 
    Sig = COV(:,:,t);
    mu = ER(:,t);
    % Determine if we need to rebalance
    if_rebal = ~isempty(find(Data.rebaldates == dates(t)));
    if if_rebal    %% We need to rebalance
        SigStale = Sig;
        if if_debug
            disp('Rebalancing');
        end
        switch param.PortConstr
            case {'equal'}
                wNew = portValCurrent .* (ones(nAssets,1)/nAssets);
            case {'equalvol'}
                vols = sqrt(diag(Sig));
                volsinv = 1./vols;
                volsinvsum = sum(volsinv);
                wNew = portValCurrent .* (volsinv/volsinvsum);
            case {'gmv'}
                % FOR YOU TO IMPLEMENT USING QUADPROG
                % Long-short
%                 wNew = portValCurrent.*quadprog(Sig,[],[],[],ones(1,size(Sig,1)),1);
%                 % Long-only
                 wNew = portValCurrent.*quadprog(Sig,[],[],[],ones(1,size(Sig,1)),1,zeros(size(Sig,1),1));
            case {'mv'}
                % FOR YOU TO IMPLEMENT USING QUADPROG
                % Here lambda stands for 2*lambda!!!
                lambda2 = 8;
                % Long-short
%                 wNew = portValCurrent.*quadprog(Sig*lambda2,-mu',[],[],ones(1,size(Sig,1)),1);
%                 % Long-only
                wNew = portValCurrent.*quadprog(lambda2*Sig,-mu',[],[],ones(1,size(Sig,1)),1,zeros(size(Sig,1),1));
        end
    else %% We don't need to rebalance
        wNew = wCurrent;
    end
    W(:,t) = wNew;
    RiskProj(t) = sqrt(wNew'*Sig*wNew)/sum(wNew);
    RiskContrProj(:,t) = wNew.*(SigStale*wNew/(wNew'*SigStale*wNew));
    %RiskContrProj(:,t) = wNew.*(Sig*wNew/(wNew'*Sig*wNew));
    trades = wNew - wCurrent;
    turnover(t) = sum(abs(trades))/portValCurrent;
    
    % Calculate portfolio at the end of the period
    % (If needed in the futute, add cash position here)
    wCurrent = wNew .* (ones(size(Rets,1),1)+Rets(:,t));
    securityRet(:,t-1) = wNew .* Rets(:,t);
    portValCurrent = sum(wCurrent);
    portVal(t) = portValCurrent;
     if portVal(t) < 0
         fprintf('Here val less than zero! %d\n',t);
         portVal(t)
         portVal(t-1)
         portVal(t-2)
         portVal(t-3)
         error('1');
     end
     portRet(t-1) = (portVal(t)-portVal(t-1))./portVal(t-1);
    retdates(t-1) = dates(t);
end

% Populate output structure
output_struct.portVal = portVal;
output_struct.dates = dates;
output_struct.W = W;
output_struct.RiskContrProj = RiskContrProj;
output_struct.RiskProj = RiskProj;
output_struct.cash = cash;
output_struct.turnover = turnover;
output_struct.secnames = Data.secnames;
output_struct.securityRet = securityRet;
output_struct.portRet = portRet;
output_struct.retdates = retdates;

return

